File: 1361.验证二叉树.js
/*
 * @lc app=leetcode.cn id=1361 lang=javascript
 *
 * [1361] 验证二叉树
 */

// @lc code=start
/**
 * @param {number} n
 * @param {number[]} leftChild
 * @param {number[]} rightChild
 * @return {boolean}
 */
var validateBinaryTreeNodes = function(n, leftChild, rightChild) {
    
    // 1. 统计入度（父节点数）
    let indegree = new Array(n).fill(0);

    for (let i = 0; i < n; i++) {

        if (leftChild[i] !== -1) {
            indegree[leftChild[i]]++;
            if (indegree[leftChild[i]] > 1) return false;
        }

        if (rightChild[i] !== -1) {
            indegree[rightChild[i]]++;
            if (indegree[rightChild[i]] > 1) return false;
        }
    }

    // 2. 找唯一根节点
    let root = -1;

    for (let i = 0; i < n; i++) {
        if (indegree[i] === 0) {
            if (root !== -1) return false; // 多个根
            root = i;
        }
    }

    if (root === -1) return false; // 没有根


    // 3. DFS检查环和连通性
    let visited = new Array(n).fill(false);

    function dfs(node) {
        if (node === -1) return true;

        if (visited[node]) return false; // 有环

        visited[node] = true;

        return dfs(leftChild[node]) &&
               dfs(rightChild[node]);
    }

    if (!dfs(root)) return false;

    // 4. 必须访问所有节点
    return visited.every(v => v === true);
};
// @lc code=end

File: 1361.验证二叉树.java
/*
 * @lc app=leetcode.cn id=1361 lang=java
 *
 * [1361] 验证二叉树
 */

// @lc code=start
class Solution {
    public boolean validateBinaryTreeNodes(int n, int[] leftChild, int[] rightChild) {
        // 1. 记录每个节点父节点数
        int[] indegree = new int[n];

        for (int i = 0; i < n; i++) {

            if (leftChild[i] != -1) {
                indegree[leftChild[i]]++;
                if (indegree[leftChild[i]] > 1) return false; // 多个父节点
            }

            if (rightChild[i] != -1) {
                indegree[rightChild[i]]++;
                if (indegree[rightChild[i]] > 1) return false;
            }
        }

        // 2. 找根节点
        int root = -1;

        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                if (root != -1) return false; // 多个根
                root = i;
            }
        }

        if (root == -1) return false; // 没有根


        // 3. DFS遍历检查连通性和环
        boolean[] visited = new boolean[n];

        if (!dfs(root, leftChild, rightChild, visited))
            return false;

        // 必须访问所有节点
        for (boolean v : visited) {
            if (!v) return false;
        }

        return true;
    }


    private boolean dfs(int node, int[] left, int[] right, boolean[] visited) {

        if (node == -1) return true;

        if (visited[node]) return false; // 出现环

        visited[node] = true;

        return dfs(left[node], left, right, visited) &&
               dfs(right[node], left, right, visited);
    }
}
// @lc code=end

